### PR TITLE
Tox regression testing

### PR DESCRIPTION
**Pull request by markrtuttle**
This pull request consists of three commits.

The first two are minor bug fixes.

The third adds a directory named test that includes a set of regression tests for most of the viewer functionality.  The regression testing is driven by the tox tool that is driven by the tox.ini init file.  tox generates the pip installation package with setup.py, the launches a sequence of virtual environments, one for each test listed in tox.ini, and installs the pip package in the fresh virtual environment and runs the test.

Note: Some existing regression tests are not run by tox.  The reason is that they cannot tolerate relocation in the file system (the location of the git repository) because of pathnames compiled into the goto binary used for the test.  This will require reimplementing the --blddir flag (the current --srcdir is the source root on the local machine, the --blddir is the source root on the machine used to build the goto binary).  Easy to do, but not yet done.

The file to focus on is ./tox.ini listed at the very end of the pull request.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+/make-property
+/make-reachable
+/make-result
+/make-source
+/make-symbol
+/make-trace
+/build
+/dist
+/.tox
+__pycache__/
```
Comment (danielsn) [inline .gitignore:21]: sort

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+[testenv:make_symbol]
+commands =
+    make -j {env:NPROC} -C tests/make-symbol clean
+    make -j {env:NPROC} -C tests/make-symbol
+
+[testenv:make_trace]
+commands =
+    make -j {env:NPROC} -C tests/make-trace clean
+    make -j {env:NPROC} -C tests/make-trace
```
Comment (danielsn) [inline tox.ini:54]: newline

File: tests/make-trace/compare-traces [hunk 3]
```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env python3
```
Comment (danielsn) [inline tests/make-trace/compare-traces:1]: licence header

File: tests/make-trace/compare-traces [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import sys
+
+
+def parse_json(name):
+    with open(name) as data:
+        return json.load(data)['viewer-trace']['traces']
+
+################################################################
+
+def compare_assign_txt_json(stxt, sjson):
```
Comment (danielsn) [inline tests/make-trace/compare-traces:15]: comment describing what this is doing

File: tests/make-trace/compare-traces [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    return True
+
+################################################################
+
+def main():
+    ttxt = parse_json(sys.argv[1])
+    tjson = parse_json(sys.argv[2])
+    txml= parse_json(sys.argv[3])
+
+    for key in ttxt.keys():
```
Comment (danielsn) [inline tests/make-trace/compare-traces:181]: xml is not called here?

File: tests/make-source/compare-sources [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+def parse_json(name):
+    with open(name) as data:
+        return json.load(data)
+
+def main():
+    build = parse_json(sys.argv[1])
+    find = parse_json(sys.argv[2])
+    walk = parse_json(sys.argv[3])
+
+    if not sources_contains(build, find):
```
Comment (danielsn) [inline tests/make-source/compare-sources:38]: maybe just create sets and check set containment

File: tests/make-result/compare-result [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    rtxt = parse_json(sys.argv[1])
+    rjson = parse_json(sys.argv[2])
+    rxml= parse_json(sys.argv[3])
+
+    if compare_results(rtxt, rjson) != 0:
+        exit(1)
+    if compare_results(rtxt, rxml) != 0:
+        exit(2)
+    if compare_results(rjson, rxml) != 0:
+        exit(3)
```
Comment (danielsn) [inline tests/make-result/compare-result:56]: different exit codes is a cute idea

File: tests/make-result/compare-result [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                continue
+            return 2
+        elif key == 'status':
+            if compare_status(first[key], second[key]):
+                continue
+            return 3
+        else:
+            if first[key] == second[key]:
+                continue
+            return 4
```
Comment (danielsn) [inline tests/make-result/compare-result:39]: magic numbers?

File: tests/make-loop/Makefile [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+FREERTOS = /Users/mrtuttle/cbmc/scripts/cbmc-viewer/make/test/freertos
+
+default: loop
+
+loop: loop-goto.json loop-json.json loop-xml.json loop-goto2.json
+	diff loop-goto.json loop-xml.json
+	diff loop-goto.json loop-json.json
+	diff loop-goto.json loop-goto2.json
+	-diff {.,output}/loop-goto.json
```
Comment (danielsn) [inline tests/make-loop/Makefile:13]: why `-` here?

**Comment by ligurio** [comment]
@markrtuttle will you continue to work on PR?

**Comment by markrtuttle** [comment]
Yes, @ligurio, I do plan to finish.

**Comment by markrtuttle** [comment]
Closing in favor of https://github.com/awslabs/aws-viewer-for-cbmc/pull/54

**Comment by markrtuttle** [comment]
Closing in favor of https://github.com/awslabs/aws-viewer-for-cbmc/pull/55

**Pull request closed without merging**